File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
from collections import deque


def is_nearby(str1, str2):
    cnt = 0
    for s1, s2 in zip(str1, str2):
        if s1 != s2:
            cnt += 1
    return True if cnt == 1 else False


def solution(begin, target, words):
    n = len(words)
    d = {}
    for w in words:
        d[w] = []
    for i in range(n):
        for j in range(i, n):
            if is_nearby(words[i], words[j]):
                d[words[i]].append(words[j])
                d[words[j]].append(words[i])
    dq = deque()
    for w in words:
        if is_nearby(begin, w):
            dq.appendleft([w, 0])
    while dq:
        word, cnt = dq.pop()
        if word == target:
            return cnt + 1
        if cnt > n:
            break
        for next_word in d[word]:
            dq.appendleft([next_word, cnt + 1])
    return 0

File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
	int n = 0;
	int m = 0;
	Map<String, List<String>> adj = new HashMap<>();
	Deque<A> dq = new LinkedList<>();
	Set<String> visited = new HashSet<>();

	public int solution(String begin, String target, String[] words) {
		n = words.length;
		m = begin.length();

		adj.put(begin, new ArrayList<>());
        for (int i = 0; i < n; i++) adj.put(words[i], new ArrayList<>());
        
		visited.add(begin);
		for (int i = 0; i < n; i++) {
			int cnt = 0;
			for (int k = 0; k < m; k++) {
				if (begin.charAt(k) != words[i].charAt(k))
					cnt++;
			}
			if (cnt == 1) adj.get(begin).add(words[i]);
		}

        boolean isTargetInWords = false;
        
		for (int i = 0; i < n; i++) {
			String cur = words[i];
            
            if (cur.equals(target)) isTargetInWords = true;
            
			// adj.put(cur, new ArrayList<>());
			for (int j = i + 1; j < n; j++) {
				int cnt = 0;
				for (int k = 0; k < m; k++) {
					if (cur.charAt(k) != words[j].charAt(k))
						cnt++;
				}

				if (cnt == 1) {
                    adj.get(cur).add(words[j]);
                    adj.get(words[j]).add(cur);
                }
                
			}
		}
        
        if (!isTargetInWords) return 0;

		dq.add(new A(begin, 0));
		while (!dq.isEmpty()) {
			A a = dq.poll();
			if (a.word.equals(target)) return a.distance;

			for (String aa: adj.get(a.word)) {
				if (!visited.contains(aa)) {
					dq.add(new A(aa, a.distance + 1));
				}
			}
		}

		return 0;
	}
}

class A {
	String word;
	int distance;

	public A( String word, int distance) {
		this.distance = distance;
		this.word = word;
	}
}